src/ex3/TypeCheckerVisitor.java

```diff
@@ -44,112 +73,278 @@ public void visit(VarDecl varDecl) {
+        }
+        whileStatement.body().accept(this);
     }
 
     @Override
     public void visit(SysoutStatement sysoutStatement) {
-
+        sysoutStatement.arg().accept(this);
+        var argType = getNodeType(sysoutStatement.arg());
+        if(!argType.equals("int")){
+            throw new SemanticException("Error: print statement expected int but got " + argType); // Point 20
+        }
     }
 
     @Override
     public void visit(AssignStatement assignStatement) {
-
+        var lvType = assignStatement.getSymbolTable().varLookup(assignStatement.lv()).getDecl();
+        assignStatement.rv().accept(this);
+        var rvType = getNodeType(assignStatement.rv());
+        if(!classMap.isValidSubclass(lvType, rvType)){
+            throw new SemanticException(String.format("Error: tried to assign %s to a %s variable", rvType, lvType)); // Point 16
+        }
     }
 
     @Override
     public void visit(AssignArrayStatement assignArrayStatement) {
+        //Point 23
+        var lvType = assignArrayStatement.getSymbolTable().varLookup(assignArrayStatement.lv()).getDecl();
+        if(!lvType.equals("int[]")){
+            throw new SemanticException("Error: Tried to make array access on a " + lvType);
+        }
+
+        assignArrayStatement.index().accept(this);
+        var indexType = getNodeType(assignArrayStatement.index());
+        if(!indexType.equals("int")){
+            throw new SemanticException("Error: Array index expecting int but got " + indexType);
+        }
+
+        assignArrayStatement.rv().accept(this);
+        var rvType = getNodeType(assignArrayStatement.rv());
+        if(!rvType.equals("int")){
+            throw new SemanticException("Error: Tried to assign " + rvType + " into an array");
+        }
 
     }
 
     @Override
     public void visit(AndExpr e) {
+        e.e1().accept(this);
+        var e1Type = getNodeType(e.e1());
+        if(!e1Type.equals("boolean")){
+            throw new SemanticException("Error: And operator expected boolean but got " + e1Type);
+        }
+
+        e.e2().accept(this);
+        var e2Type = getNodeType(e.e2());
+        if(!e2Type.equals("boolean")){
+            throw new SemanticException("Error: And operator expected boolean but got " + e2Type);
+        }
+
+        addNodeType(e, "boolean");
 
     }
 
     @Override
     public void visit(LtExpr e) {
+        e.e1().accept(this);
+        var e1Type = getNodeType(e.e1());
+        if(!e1Type.equals("int")){
+            throw new SemanticException("Error: LT operator expected int but got " + e1Type);
+        }
+
+        e.e2().accept(this);
+        var e2Type = getNodeType(e.e2());
+        if(!e2Type.equals("int")){
+            throw new SemanticException("Error: LT operator expected int but got " + e2Type);
+        }
+
+        addNodeType(e, "boolean");
 
     }
 
     @Override
     public void visit(AddExpr e) {
+        e.e1().accept(this);
```

Comment:
maybe create a helper method to do this instead of copy pasting in lt, add, sub, mult

Reply:
Done